modules/mid_registrar/save.c

```diff
@@ -1348,7 +1345,7 @@ static inline int save_restore_rpl_contacts(struct sip_msg *req,
 			goto update_usrloc;
 
 		calc_ob_contact_expires(rpl, _c->expires, &e_out, 0);
-		e_out -= get_act_time();
+		e_out = e_out == 0 ? 0 : e_out - get_act_time();
```

Comment:
The original code seems to have been written with the assumption that a 200 OK from the main registrar **cannot ever** include a Contact with **0** expiration (if a contact is included, it MUST have a non-zero expiry).  While it is a reasonable assumption to make, it makes the code very fragile to buggy or malicious SIP traffic, so your fix is excellent.

Reply:
This was REGISTER from UA (to unregister):
```
REGISTER sip:test.voip.tel SIP/2.0
Via: SIP/2.0/UDP 52.10.16.16:64254;rport;branch=z9hG4bKPjEh7mH2LlAI5APPECJvR03S2YSqAYTNKi
Max-Forwards: 69
From: "new_test" <sip:102_test@test.voip.tel>;tag=1ZFs-hb.820fBtFb3b37tiygGdaIv2Qo
To: "new_test" <sip:102_test@test.voip.tel>
Call-ID: GHTV.sGO8eD1GrapKMKGAVna3fyROY2d
CSeq: 24771 REGISTER
User-Agent: Telephone 1.5.2
Contact: <sip:102_test@52.10.16.16:64254;ob>
Expires: 0
Content-Length:  0
```
Then main registrar sent 200 OK (its' behavior can be wrong, as it contains Contact with `expires=0`):
```
SIP/2.0 200 OK
Record-Route: <sip:172.18.16.241;r2=on;lr=on>,...
User-Agent: VoIP
Contact: <sip:102_test@172.26.26.231:5060;ctid=2356649242509991926;realip=52.10.16.16:64254;natip=52.10.16.16:64254>;expires=0
CSeq: 24771 REGISTER
Call-ID: GHTV.sGO8eD1GrapKMKGAVna3fyROY2d
Expires: 0
To: "new_test" <sip:102_test@test.voip.tel>
From: "new_test" <sip:102_test@test.voip.tel>;tag=1ZFs-hb.820fBtFb3b37tiygGdaIv2Qo
Content-Length: 0
Via: SIP/2.0/UDP 172.16.16.231:5060;branch=z9hG4bK904a.c3614d1.0
Via: SIP/2.0/UDP 52.10.16.16:64254;received=52.10.16.16;rport=64254;branch=z9hG4bKPjEh7mH2LlAI5APPECJvR03S2YSqAYTNKi
```
So mid_registrar transforms this reply to:
```
SIP/2.0 200 OK
Record-Route: <sip:172.18.16.241;r2=on;lr=on>,...
User-Agent: VoIP
CSeq: 24771 REGISTER
Call-ID: GHTV.sGO8eD1GrapKMKGAVna3fyROY2d
X-ID: hx-d82d416981dfd86c353434847c223082
To: "new_test" <sip:102_test@test.voip.tel>
From: "new_test" <sip:102_test@test.voip.tel>;tag=1ZFs-hb.820fBtFb3b37tiygGdaIv2Qo
Content-Length: 0
Via: SIP/2.0/UDP 52.10.16.16:64254;received=52.10.16.16;rport=64254;branch=z9hG4bKPjEh7mH2LlAI5APPECJvR03S2YSqAYTNKi
Contact: <sip:102_test@52.10.16.16:64254;ob>;expires=18446744072071357650;received="sip:52.10.16.16:64254"
```

Reply:
Perfect!  Exactly as I imagined: the main-registrar replies with a 200 OK advertising a Contact with `;expires=0`:

* does this make sense?  Not really, as a `0` expiry means "this contact is valid for 0 more seconds".  So why advertise something that does not exist?
* is this a common occurrence?  Probably, and the mid-registrar implementation should be equipped to deal with any corner-cases on the downstream device without reaching a bad internal state.

Thank you for all your help, I can safely backport to `3.1`+ now.  And we are done here! :-)